### PR TITLE
Re-enable flaky logging tests, add logs to output redirection providers, fix potential race condition in closing pipe output.

### PR DESCRIPTION
**Pull request by jkotalik**
For #897 . Added more logging. Also, the FileOutputManager's implementation has changed since disabling the test. I keep an eye on any failures, looking for the cause.

### CODE REVIEW

File: src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp [hunk 1]
```diff
@@ -175,19 +163,19 @@ FileOutputManager::Start()
 
     if (!SetStdHandle(STD_OUTPUT_HANDLE, m_hLogFileHandle))
     {
-        hr = HRESULT_FROM_WIN32(GetLastError());
-        return hr;
+        return LOG_IF_FAILED(HRESULT_FROM_WIN32(GetLastError()));
     }
 
     if (!SetStdHandle(STD_ERROR_HANDLE, m_hLogFileHandle))
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/FileOutputManager.cpp]: RETURN_LAST_ERROR_IF

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 2]
```diff
@@ -126,12 +122,10 @@ HRESULT PipeOutputManager::Start()
 
     if (m_hErrThread == NULL)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: RETURN_LAST_ERROR_IF_NULL

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 3]
```diff
@@ -2,19 +2,22 @@
...
-    m_hErrThread(INVALID_HANDLE_VALUE),
-    m_fDisposed(FALSE)
+    m_hErrThread(INVALID_HANDLE_VALUE)
 {
 }
 
 PipeOutputManager::~PipeOutputManager()
 {
-    StopOutputRedirection();
+    // For some reason, we can't put StopOutputRedirection as the argument here.
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: This doesn't work. If you call StopOutputRedirection destructor would call it again.

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp (hunk 2 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: Actually this check is wrong. It should be checking for invalid handle.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: nvrm I'm wrong.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-createthread#return-value

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp (hunk 3 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: I see. Two things I can do:
- Make StopOutputRedirection private
- StopOutputRedirection just does a double check lock to set a bool.

I'll do the second one for now.

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 4]
```diff
@@ -49,6 +57,7 @@ PipeOutputManager::StopOutputRedirection()
             // if the thread is still running, we need kill it first before exit to avoid AV
             if (GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: wrap all calls here into logging macroses

Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: You referring to LOG_IF_FAILED()?

Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: LOG_LAST_ERROR_IF

Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: For things that return bool and set LastError

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 5]
```diff
@@ -40,15 +49,16 @@ PipeOutputManager::StopOutputRedirection()
...
+        LOG_LAST_ERROR_IF(GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0) &&
         dwThreadStatus == STILL_ACTIVE)
     {
         // wait for gracefullshut down, i.e., the exit of the background thread or timeout
-        if (WaitForSingleObject(m_hErrThread, PIPE_OUTPUT_THREAD_TIMEOUT) != WAIT_OBJECT_0)
+        if (LOG_LAST_ERROR_IF(WaitForSingleObject(m_hErrThread, PIPE_OUTPUT_THREAD_TIMEOUT) != WAIT_OBJECT_0))
         {
             // if the thread is still running, we need kill it first before exit to avoid AV
-            if (GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+            if (LOG_LAST_ERROR_IF(GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE))
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: Just wrap calls into LOG_LAST_ERROR_IF, not entire expressions.

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 6]
```diff
@@ -40,15 +49,17 @@ PipeOutputManager::StopOutputRedirection()
 
     if (m_hErrThread != NULL &&
         m_hErrThread != INVALID_HANDLE_VALUE &&
-        GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 &&
+        LOG_LAST_ERROR_IF(GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0) &&
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: You need to log when result is 0

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 7]
```diff
@@ -95,21 +107,18 @@ HRESULT PipeOutputManager::Start()
...
 
     // TODO this still doesn't redirect calls in native, like wprintf
     if (!SetStdHandle(STD_ERROR_HANDLE, hStdErrWritePipe))
     {
-        hr = HRESULT_FROM_WIN32(GetLastError());
-        goto Finished;
+        return LOG_IF_FAILED(HRESULT_FROM_WIN32(GetLastError()));
     }
 
     if (!SetStdHandle(STD_OUTPUT_HANDLE, hStdErrWritePipe))
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: RETURN_LAST_ERROR_IF

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp (hunk 6 above)
Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: Yeah, these checks are incorrect right now.

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 8]
```diff
@@ -38,17 +47,20 @@ PipeOutputManager::StopOutputRedirection()
...
-        // wait for gracefullshut down, i.e., the exit of the background thread or timeout
+        // wait for graceful shutdown, i.e., the exit of the background thread or timeout
         if (WaitForSingleObject(m_hErrThread, PIPE_OUTPUT_THREAD_TIMEOUT) != WAIT_OBJECT_0)
         {
             // if the thread is still running, we need kill it first before exit to avoid AV
-            if (GetExitCodeThread(m_hErrThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+            if (!LOG_LAST_ERROR_IF(GetExitCodeThread(m_hErrThread, &dwThreadStatus) == 0) &&
+                dwThreadStatus == STILL_ACTIVE)
             {
+                WLOG_WARNF(L"Thread reading stdout/err hit timeout, forcibly closing thread.");
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: `LOG_WARN`

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 9]
```diff
@@ -68,11 +80,13 @@ PipeOutputManager::StopOutputRedirection()
     if (m_fdPreviousStdOut != -1)
     {
         _dup2(m_fdPreviousStdOut, _fileno(stdout));
+        WLOG_INFOF(L"Restoring original stdout of stdout: %d", m_fdPreviousStdOut);
     }
 
     if (m_fdPreviousStdErr != -1)
     {
         _dup2(m_fdPreviousStdErr, _fileno(stderr));
+        WLOG_INFOF(L"Restoring original stdout of stderr: %d", m_fdPreviousStdErr);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: LOG_INFOF

File: src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp [hunk 10]
```diff
@@ -68,11 +80,13 @@ PipeOutputManager::StopOutputRedirection()
     if (m_fdPreviousStdOut != -1)
     {
         _dup2(m_fdPreviousStdOut, _fileno(stdout));
+        WLOG_INFOF(L"Restoring original stdout of stdout: %d", m_fdPreviousStdOut);
     }
 
     if (m_fdPreviousStdErr != -1)
     {
         _dup2(m_fdPreviousStdErr, _fileno(stderr));
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/PipeOutputManager.cpp]: Are you going to do all _dup2 in other PR?

**Pull request merged**